wgpu

jazzfool opened the pull request.

Commit:
preliminary work on wgpu

Commit:
enable msaa ctor for canvas

Commit:
samplers and shaders

aleokdev commented on src/graphics/canvas.rs:
```diff
@@ -1,392 +1,85 @@
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum CanvasLoadOp {
+    /// Keep the existing contents of the image.
+    DontClear,
```
DoNotClear is maybe better, Dont sounds too informal

jazzfool replied:
i think this language is ok, vulkan for example has `VK_ATTACHMENT_LOAD_OP_DONT_CARE`

Commit:
initial work on drawing code

PSteinhaus committed:
removed unnecessary &mut Context;
removed/replaced unnecessary duration conversions;

PSteinhaus committed:
replaced usages of removed timer functions with standard duration functions

Commit:
working wgpu renderer

aleokdev commented on README.md:
```diff
@@ -1,188 +1,40 @@
 
- * Open an issue on [the Github issue tracker](https://github.com/ggez/ggez/issues)
- * Say hi on the [unofficial Rust Discord server](http://bit.ly/rust-community), the [Rust Gamedev server](https://discord.gg/yNtPTb2) or the [good-web-game Discord server](https://discord.gg/jum3Fjek2A)
+The following platforms are untested, but the backend is compatible (i.e. no major changes needed to make fully compatible):
```
What about web?

jazzfool replied:
ah yeah, web too

Commit:
add text drawing

aleokdev commented on src/error.rs:
```diff
@@ -5,7 +5,7 @@ use std::sync::Arc;
 
 /// An enum containing all kinds of game framework errors.
-#[derive(Error, Debug, Clone)]
+#[derive(Error, Debug)]
```
No clone?

jazzfool replied:
spawnerror
might `Arc` it later to restore clone tho

aleokdev commented on src/error.rs:
```diff
@@ -54,6 +54,9 @@ pub enum GameError {
     /// Something went wrong with the `lyon` shape-tesselation library.
     #[error("lyon tesellation error: {0}")]
     LyonError(String),
+    /// Something went wrong when spawing a text with `futures`.
```
spawning*

aleokdev commented on src/graphics/canvas.rs:
```diff
@@ -538,8 +565,105 @@ impl<'a> Canvas<'a> {
+                .draw_queued_with_transform(
+                    device,
+                    staging_belt,
+                    unsafe { encoder.as_mut().unwrap() },
```
Missing safety guarantee

jazzfool replied:
pass is the only thing borrowing encoder so that's explicitly dropped above
quite harmless anywho

Commit:
port meshbuilder

aleokdev committed:
Add `SpriteBatch::image()`

nobbele committed:
Merge pull request #1017 from aleokdev/spritebatch-img

Add `SpriteBatch::image()`

Commit:
port snake example

Commit:
streamline api, impl drawparam origin

aleokdev commented on src/audio.rs:
```diff
@@ -85,9 +81,9 @@ pub struct SoundData(Arc<[u8]>);
 impl SoundData {
     /// Load the file at the given path and create a new `SoundData` from it.
-    pub fn new<P: AsRef<path::Path>>(context: &mut Context, path: P) -> GameResult<Self> {
+    pub fn new<P: AsRef<path::Path>>(fs: &Filesystem, path: P) -> GameResult<Self> {
```
At this point this PR is really just changing everything

aleokdev commented:
Should we refcount image? wasn't it refcounted before? Canvas uses a ref to it instead of an arc

Commit:
fix draw transforms (ex. 03_drawing)

PSteinhaus committed:
re-added `apply_transformations` but renamed it to `apply_projection` as that's all it does now;
it simply calls `reset_global_mvp`, which is has the same effect as `set_global_mvp(IDENTITY)` and is necessary to reliably apply custom projection matrices set with `set_projection` or `mul_projection`;

Commit:
- new text renderer
- growing uniform buffer
- reworked canvas internals

Commit:
make text use Cow

Commit:
fix example 01

Commit:
- fix glyph cache uploading
- add frame canvas QOL ctor
- update examples

Commit:
Merge remote-tracking branch 'upstream/devel' into wgpu

aleokdev committed:
Refactor gamepad and keyboard ctx

aleokdev committed:
Refactor timer context

aleokdev committed:
Rename context members

PSteinhaus commented:
I'm still trying to read and understand more of your fork. Until I get a deeper understanding I can only be of limited help and point you towards tiny things, such as:

- Why do you use enumerate here, if you don't use the enumeration? It can just be removed, right? https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/gpu/growing.rs#L27
- These comments should probably refer to text, not meshes, right? https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/canvas.rs#L282 https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/canvas.rs#L296 https://github.com/aleokdev/ggez/blob/757de08f14f29652a060ded228938519e119349c/src/graphics/canvas.rs#L310

Commit:
remove unnecessary enumerate

Commit:
correct text shader docs

aleokdev committed:
Rename `timer` member to `time`; Fix examples

aleokdev committed:
Feature-gate audio and gamepad

aleokdev committed:
Fix warning

aleokdev committed:
Refactor + move input contexts to InputContext

Commit:
reexport graphics submodules

Commit:
remove unused gfx type

Commit:
canvas docs and QOL

Commit:
simplify archandle ID

Commit:
clarify instancearray doc

Commit:
respect vsync conf

jazzfool commented:
TODO list before this can be considered for a PR
- [x] Reach API parity with `ggez::graphics`
- [x] Transition all examples
- [x] Publicly expose WGPU types (allows for user integration of other rendering, e.g. egui)
    - Also involves exposing the current frame's `CommandEncoder`
- [x] Stress test *all* parts of the new API to find bugs
    - [x] Growing uniform buffer
    - [x] MSAA canvases
    - [x] Instance arrays
    - [x] Custom shaders
    - [x] Custom shader uniforms, textures, and samplers
- [x] ? Z-sorted queue for `Canvas`
- [ ] ~~? Test web support~~ out of scope for PR

Commit:
improved canvas error handling

Commit:
mesh constructor helpers

Commit:
impl image readback and encoding

Commit:
expose internal frame image

Commit:
fix examples

Commit:
port animation example

Commit:
Merge branch 'wgpu' of https://github.com/alexdevteam/ggez into wgpu

Commit:
impl canvas blend modes

Commit:
fix instanced rendering

Commit:
port bunnymark example

Commit:
restore original readme

Commit:
fix struct alignment

Commit:
restore original drawparam

Commit:
port canvas_subframe example

Commit:
fix non-linear srgb colors being sent in uniforms

Commit:
port colorspace example

Commit:
change how drawparam image scaling works

Commit:
expose wgpu internal types

Commit:
make linearcolor public

Commit:
port cube example

Commit:
make begin/end_frame public

Commit:
port eventloop example

Commit:
port files example

Commit:
window API parity

Commit:
make flush_text private again

Commit:
- port shadows example
- add blurring to shadows example

aleokdev committed:
Update deprecation warnings

Commit:
respect window setup msaa

Commit:
port graphics_settings example

Commit:
projection matrix API parity

Commit:
impl projection() and mul_projection()

Commit:
impl z-sorted canvas

Commit:
port examples to new canvas

Commit:
- port remaining examples
- remove irrelevant examples

aleokdev committed:
Refactor filesystem + Rename ctx member to fs

aleokdev committed:
Update deprecation messages

aleokdev committed:
Remove `InputContext`

aleokdev committed:
Fix no default features cfg

aleokdev committed:
Revert removing event::{Axis, Button}

Commit:
add convenience accessors to mesh

Commit:
make instancearray len arc/atomic

PSteinhaus committed:
changed more `RodioAudioDevice` mentions to `AudioDevice`

PSteinhaus committed:
Merge pull request #1019 from aleokdev/newctx

Use subcontexts directly instead of free functions

Commit:
add support for text rotation

aleokdev committed:
Merge remote-tracking branch 'upstream/devel' into wgpu

aleokdev committed:
Revert some unrelated changes + Remove thiserror

aleokdev committed:
Run `cargo fmt`

aleokdev committed:
Revert more unrelated changes

aleokdev committed:
Fix most clippy warnings

Commit:
add explicit error for to_pixels on MSAA images.
wgpu should already throw an error when this happens, but this should be more user friendly

Commit:
fix cargo clippy messages

Commit:
update readme

Commit:
fix 05_astroblasto example screenshotting

Commit:
split ctx in image::from_path

Commit:
fix cargo clippy messages for examples

Commit:
revert readme formatting

aleokdev commented:
Refer to the upstream PR
